libflux/src/semantic/types.rs

```diff
@@ -116,27 +205,251 @@ impl fmt::Display for MonoType {
+                })
+            }
+        }
+    }
+
+    fn contains(&self, tv: Tvar) -> bool {
+        match self {
+            MonoType::Bool
+            | MonoType::Int
+            | MonoType::Uint
+            | MonoType::Float
+            | MonoType::String
+            | MonoType::Duration
+            | MonoType::Time
+            | MonoType::Regexp => false,
+            MonoType::Var(tvr) => tv == *tvr,
+            MonoType::Arr(arr) => arr.contains(tv),
+            MonoType::Row(row) => row.contains(tv),
+            MonoType::Fun(fun) => fun.contains(tv),
+        }
+    }
+}
+
 // Tvar stands for type variable.
 // A type variable holds an unknown type.
-#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
-pub struct Tvar(i64);
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct Tvar(pub i64);
 
 impl fmt::Display for Tvar {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(f, "t{}", self.0)
     }
 }
 
+impl Tvar {
+    fn apply(self, sub: &Subst) -> MonoType {
+        match sub.lookup(self) {
+            Some(t) => t.clone(),
+            None => MonoType::Var(self),
+        }
+    }
+
+    fn unify(self, with: MonoType, cons: Const) -> Result {
+        match with {
+            MonoType::Var(tv) => {
+                if self == tv {
+                    // The empty substitution will always
+                    // unify a type variable with itself.
+                    Ok((Subst::empty(), cons))
+                } else {
+                    // Unify two distinct type variables.
+                    // This will update the kind constraints
+                    // associated with these type variables.
+                    self.unify_with_tvar(tv, cons)
+                }
+            }
+            _ => {
+                if with.contains(self) {
+                    // Invalid recursive type
+                    Err(Error::Occurs { tv: self, t: with })
+                } else {
+                    // Unify a type variable with a monotype.
+                    // The monotype must satisify any
+                    // constraints placed on the type variable.
+                    self.unify_with_type(with, cons)
+                }
+            }
+        }
+    }
+
+    fn unify_with_tvar(self, tv: Tvar, cons: Const) -> Result {
+        let mut cons = cons;
+        // Gather the kind constraints for both type variables
+        let kinds: HashSet<Kind> = cons
+            .0
+            .get(&self)
+            .unwrap_or(&HashSet::new())
+            .union(cons.0.get(&tv).unwrap_or(&HashSet::new()))
+            .map(|&kind| kind)
+            .collect();
+        let sub = Subst::init(maplit::hashmap! {self => MonoType::Var(tv)});
+        if kinds.len() > 0 {
+            // Update the kind constraints
+            cons.0.remove(&self);
+            cons.0.remove(&tv);
+            cons.0.insert(tv, kinds);
```

Comment:
insert will replace the existing value, why do we remove it first?

Reply:
Typo